### DISCUSSION TITLE
How to tell in a library function if OpENer is running?

### DISCUSSION THREAD

**Discussion by shankar2k** [EIPStackGroup/OpENer#411] [Q&A]
I am trying to use OpENer as a shared library in order to access it from LabVIEW.  When I call an OpENer function from LabVIEW, the shared library is loaded into memory, and all of the global variables can be potentially accessed. Is there a variable to check or an OpENer function that I can call that will tell me if the OpENer stack is currently running? I need this so that, for example, if LabVIEW tries to get assembly instance data before the OpENer stack is running, it will be able to return an error instead of just crashing.

**Comment by MartinMelikMerkumians** [comment]
I get what you need, but not sure if OpENer can provide this. How can you tell if you can access the indicator then?

**Reply by shankar2k** [reply]
So for example, I have the following code snippet in my OpENer shared library source:

```c
char g_name_buf[HOST_NAME_MAX];

void OpLVGetHostName() {
  int rc = gethostname(g_name_buf, sizeof g_name_buf);
  g_name_buf[HOST_NAME_MAX - 1] = '\0'; /* Ensure termination */
  if(0 == rc) {
    SetCipStringByCstr(&g_tcpip.hostname, g_name_buf);
  }
}

int MyGetHostName(char* hostname) {
  if (g_tcpip.hostname.string == NULL) {
    OpLVGetHostName();
  }
  memcpy(hostname, g_tcpip.hostname.string , g_tcpip.hostname.length);
  return 0;
}
```

Even if the OpENer stack isn't running, I can access the `MyGetHostName `function from LabVIEW using the [Import Shared Library Wizard](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA03q000000YGZjCAO). This is because when the shared library is loaded, the global variable `g_tcpip` is accessible.

I am looking for a variable or a function call that is true if and only if the OpENer stack is running. 

The first idea I had was to check the value of the global variable which holds the connection id. That turns out to be `g_incarnation_id`, but in the default OpENer implementation, it is set to a random number when the OpENer stack is running, and is uninitialized and thus unspecified otherwise.

Another idea I had was to check if a certain needed class was present. For example, does the function call `GetCipClass(kCipMessageRouterClassCode)` return non-`NULL` if and only if the OpENer stack is running?

**Reply by MartinMelikMerkumians** [reply]
not sure for this one, but each ENIP device needs an identity object to identify itself
so if there is no instance 1 instance object of the identity object class, you can assume OpENer/the lib is not operational

**Reply by shankar2k** [reply]
Sorry for the delay in responding. I solved my problem by creating my own global variable `g_stack_running`, and then I turn it on and off in my version of `executeEventLoop`:

```c
bool g_stack_running = false;

bool OpLVIsStackRunning() {
  return g_stack_running;
}

static void *OpLVExecuteEventLoop(void *pthread_arg) {
  static int pthread_dummy_ret;
  (void) pthread_arg;

  g_stack_running = true;
  while(!g_end_stack) {
    if(kEipStatusOk != NetworkHandlerProcessCyclic() ) {
      OPENER_TRACE_ERR("Error in NetworkHandler loop! Exiting OpENer!\n");
      break;
    }
  }
  g_stack_running = false;

  return &pthread_dummy_ret;
}
```
